### PR TITLE
[KernelGen][Nvidia] Add linalg_cholesky operator with Triton kernel

### PR DESCRIPTION
**Pull request by XDYuanzhuLee**
## Summary
Adds a Triton kernel for `linalg_cholesky`. Computes the Cholesky decomposition of a symmetric positive-definite matrix A
or for batches of symmetric positive-definite matrices.

## Testing
- Validated against reference on device via `to_reference(inp, True)`
- Tested on: Nvidia, Tianshu, Muxi, Ascend, Hygon

## Performance
Test command: `pytest benchmark/test_linalg_cholesky.py --level core` (NVIDIA CI)

### linalg_cholesky

| dtype | Size | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|------|--------------------|-------------------|---------|
| float32 | [2, 2] | 0.110592 | 0.164864 | 0.671 |
| float32 | [4, 4] | 0.111616 | 0.161792 | 0.690 |
| float32 | [8, 8] | 0.110592 | 0.161792 | 0.684 |
| float32 | [16, 16] | 0.111616 | 0.162816 | 0.686 |
| float32 | [32, 32] | 0.111616 | 0.163840 | 0.681 |
| float32 | [64, 64] | 0.110592 | 0.162816 | 0.679 |
| float32 | [128, 128] | 0.142336 | 0.193536 | 0.735 |
| float32 | [256, 256] | 0.215040 | 0.265216 | 0.811 |

| Operator | Arithmetic Mean Speedup |
|----------|------------------------|
| linalg_cholesky | **0.705** |

## Multi-backend Testing
| Backend | Accuracy Test | Speedup (mean) | Notes |
|---|---|---|---|
| Nvidia (CI) | PASS (8 cases) | 0.705 | Primary |
| Tianshu | FAIL | 0.680 | RuntimeError: gemm of double is not supported on CoreX. |
| Muxi | PASS | 0.999 | — |
| Ascend | FAIL | — | AssertionError: Tensor-likes are not close! |
| Hygon | PASS | 13.957 | — |

## Files Changed
- `src/flag_gems/ops/linalg_cholesky.py`: Triton kernel implementation
- `tests/test_linalg_cholesky.py`: Accuracy test
- `benchmark/test_linalg_cholesky.py`: Performance benchmark
- `src/flag_gems/ops/__init__.py`: Register import and `__all__`
- `src/flag_gems/__init__.py`: Register to `_FULL_CONFIG`
- `conf/operators.yaml`: Add operator entry (kind: BLAS, stage: alpha 5.1)

### CODE REVIEW

**Review by Dongxu-H — CHANGES_REQUESTED** [review]
## Code Review Results

### 🔴 Must Fix

**1. [`src/flag_gems/ops/__init__.py`](https://github.com/flagos-ai/FlagGems/pull/3894/files#diff-xxx) - `__all__` list sort order issue**

```python
    "celu_",
+   "linalg_cholesky",
    "clamp",
```

`linalg_cholesky` starts with `l` but was added in the `c` section! It should be in the `l` section, between `lift_fresh_copy` and `linspace`:

```python
    "lift_fresh_copy",
+   "linalg_cholesky",
    "linspace",
```

---

### 🟡 Suggestions

**2. [`src/flag_gems/ops/linalg_cholesky.py:17-21`](https://github.com/flagos-ai/FlagGems/pull/3894/files#diff-xxx) - Potential stride calculation issue**

```python
A_sym = sym_like(A)
L_sym = sym_like(L)
stride_a = A_sym.stride(0)
stride_l = L_sym.stride(0)
```

For a 3D view reshaped from higher dimensions, `stride(0)` might not correctly represent the row stride. Consider using `A.stride(-2)` or `A.stride(-1)` for clarity, or verify that the symbolic stride matches the expected `n`.

**3. [`src/flag_gems/ops/linalg_cholesky.py:46`](https://github.com/flagos-ai/FlagGems/pull/3894/files#diff-xxx) - Sequential kernel for each batch**

```python
pid = tl.program_id(axis=0)
```

Each program handles one batch element. For large batch sizes, this could be inefficient. Consider:
- Adding a comment about the trade-off (sequential vs parallel)
- Or exploring a parallel algorithm for better scalability

**4. [`tests/test_linalg_cholesky.py`](https://github.com/flagos-ai/FlagGems/pull/3894/files#diff-xxx) - Missing test coverage**

Consider adding tests for:
- **Batched inputs**: `(2, 4, 4)`, `(3, 8, 8)` (multi-batch matrices)
- **float64 dtype**: PyTorch supports float64 for Cholesky
- **Edge cases**: 1x1 matrix, near-singular matrices

---

### ✅ Verified Correct

| Check | Status |
|-------|--------|
| Test coverage | ✅ unit test + benchmark exist |
| Config consistency | ✅ `operators.yaml` sorted correctly (after `layer_norm`, before `linspace`) |
| Kernel registration | ✅ `__init__.py` sorted correctly (after `lift_fresh_copy`, before `linalg_vector_norm`) |
| Naming conventions | ✅ Uses `cholesky` (correct ATen op name) |
| Cross-platform | ✅ Pure Triton, no `is_cuda` usage |
| Special cases | ✅ `upper` parameter supported, tested |

---

### Summary

The implementation has a sorting issue in `ops/__init__.py` that must be fixed. Other suggestions are for code quality and test coverage improvements.

| Issue | Severity |
|-------|----------|
| `__all__` sort order | 🔴 Must fix |
| Stride calculation | 🟡 Suggestion |
| Sequential kernel design | 🟡 Suggestion |
| Test coverage | 🟡 Suggestion |

Please fix the sort order issue before merging.

**Review by XDYuanzhuLee — COMMENTED** [review]
Thanks for the review. The `__all__` sort order has been fixed — `linalg_cholesky` moved to the `l` section between `lift_fresh_copy_out` and `linspace`.

Regarding the suggestions:
- **Stride calculation**: The strides come from `sym_like()` which preserves the actual tensor layout; this is the standard pattern used by other decomposition kernels in the repo.
- **Sequential kernel**: Good point — added in the design trade-off as a known limitation of the current Triton kernel approach for batched Cholesky.
- **Test coverage**: The current test covers the core float32 path; additional dtype/batch tests can be added in a follow-up PR.

CI re-run pending.

**Comment by bin913** [comment]
@XDYuanzhuLee Please resolve conflicts

**Review by Dongxu-H — COMMENTED** [review]
## FlagGems Operator Contribution Review

**Operator**: `linalg_cholesky`  
**PR**: #3894

### 📊 Change Statistics
| Metric | Value |
|--------|-------|
| Additions | 266 |
| Deletions | 0 |
| Files Changed | 6 |

### 📋 22-Point Checklist

| # | Requirement | Status |
|---|-------------|--------|
| 1 | Implementation file path correctness | ✅ Correct: `src/flag_gems/ops/linalg_cholesky.py` |
| 2 | KernelGen attribution in header | ✅ Present |
| 3 | Import order (stdlib → third-party → internal) | ⚠️ Could not verify |
| 4 | Function signature alignment with PyTorch API | ⚠️ Needs manual verification |
| 5 | Type support documentation | ✅ Type handling present |
| 6 | Triton kernel decorators (`@triton.jit`, etc.) | ⚠️ `@triton.jit` present, verify other decorators |
| 7 | Test file existence | ✅ Present: `tests/test_linalg_cholesky.py` |
| 8 | Test coverage (shapes, dtypes) | ✅ Shape and dtype tests present |
| 9 | Benchmark file existence | ✅ Present: `benchmark/test_linalg_cholesky.py` |
| 10 | Benchmark implementation correctness | ✅ Benchmark implemented |
| 11 | Config registration (`operators.yaml`) | ✅ Registered in `operators.yaml` |
| 12 | Module registration (`__init__.py`) | ✅ Registered in `__init__.py` |
| 13 | Ops module registration (`ops/__init__.py`) | ✅ Registered in `ops/__init__.py` |
| 14 | Logger namespace (`flag_gems.*`) | ⚠️ Missing `flag_gems.` prefix |
| 15 | Error handling | ✅ Error handling present |
| 16 | Numerical stability patterns | ✅ Stability patterns found |
| 17 | Memory efficiency | ⚠️ Needs profiling |
| 18 | Edge case tests | ⚠️ Needs verification |
| 19 | Backend support (Metax) | ❌ Not implemented |
| 20 | Documentation/Docstrings | ✅ Docstrings present |
| 21 | Code style consistency | ⚠️ Run linter manually |
| 22 | Integration verification | ⚠️ Verify end-to-end |

### 📈 Summary

| Category | Count |
|----------|-------|
| ✅ Pass | 13 |
| ⚠️ Warning | 8 |
| ❌ Fail | 1 |

**Overall**: ⚠️ Needs Attention

---

*This review was generated by an automated script following the FlagGems Operator Contribution Guidelines.*
*For detailed code review, please consult a domain expert.*

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册
- 💡 tl.load建议添加mask
- 💡 tl.store建议添加mask

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册
- 💡 tl.load建议添加mask
- 💡 tl.store建议添加mask

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册
- 💡 tl.load建议添加mask
- 💡 tl.store建议添加mask

**Review by Dongxu-H — CHANGES_REQUESTED** [review]
## Code Review: `linalg_cholesky` Implementation

感谢您实现 Cholesky 分解！这是一个重要的线性代数操作。以下是代码审查意见：

### 主要问题

**1. Kernel 实现效率问题**

当前使用嵌套循环实现 Cholesky 分解：
```python
for i in range(N):
    for j in range(i + 1):
        # 内部计算...
```
这种实现方式在 GPU 上效率很低，因为：
- Cholesky 分解具有数据依赖性，第 i 行的计算依赖于第 0 到 i-1 行的结果
- 每个 program 只处理一个矩阵，并行度受限
- 大量的内存访问和同步开销

**建议**：考虑使用以下优化方法之一：
- 分块 Cholesky 算法，利用 shared memory
- 使用 cuSOLVER 或 cuBLAS 的批量 Cholesky
- 参考 [this paper](https://arxiv.org/abs/2009.01660) 中的 GPU Cholesky 实现

**2. 对称化处理可能不正确**

```python
A_sym = (A + A.t().conj()) / 2
```
虽然这确保了对称性，但 Cholesky 分解要求输入必须是正定的。添加 eps 到对角线可以改善数值稳定性，但可能不足以处理所有情况。

**3. 缺少错误处理**

如果输入矩阵不是正定的，Cholesky 分解会失败。建议添加：
- 检测输入是否为正定矩阵
- 提供有意义的错误信息

**4. Upper 参数处理**

```python
if upper:
    L = L.transpose(-2, -1).conj()
```
这个处理是正确的，但需要确保 PyTorch 的 `aten.cholesky` 调用时传入的 upper 参数正确传递。

### 代码质量问题

**1. 导入顺序**

建议按照 PEP8 排序导入：
```python
import torch
import triton
import triton.language as tl
```

**2. 测试覆盖**

建议添加：
- 批量矩阵测试（batch dimension）
- 边界情况：1x1 矩阵、奇异矩阵处理
- 数值精度验证（与 numpy/scipy 对比）

**3. Benchmark 形状**

当前只测试了 float32，建议添加 float64 测试以验证精度。

### 总结

这是一个好的开始，但 kernel 实现需要重新设计以提高效率。对于小矩阵可能还可以接受，但对于大矩阵性能会很差。

建议先调研现有的高效 Cholesky 实现（如 CUTLASS、cuSOLVER），或者考虑使用 PyTorch 内置函数作为 fallback。

期待您的更新！

**Comment by XDYuanzhuLee** [comment]
Thanks. Conflicts resolved by merging origin/master (HEAD now at e94c13f7c). All checks pass: git diff --check clean, py_compile OK, check_operator.py --strict: 0 errors, 0 warnings.

**Review by Dongxu-H — CHANGES_REQUESTED** [review]
## 🔄 Re-Review - PR #3894

### 📊 Current Status

| Field | Value |
|-------|-------|
| `reviewDecision` | **CHANGES_REQUESTED** (Dongxu-H × 2) |
| `mergeable` | **MERGEABLE** ✅ |
| CI | python-op PASS ✅ |

**Previous reviews by Dongxu-H raised valid concerns that remain unaddressed.**

---

### 🔴 CRITICAL: Performance Regression - 30% Slower!

**Mean Speedup: 0.705x** (比 PyTorch **慢 30%！**)

| dtype | Size | Torch (ms) | Gems (ms) | Speedup |
|-------|------|------------|-----------|---------|
| float32 | [2, 2] | 0.110 | 0.165 | **0.671** ❌ |
| float32 | [64, 64] | 0.110 | 0.163 | **0.679** ❌ |
| float32 | [256, 256] | 0.215 | 0.265 | **0.811** ❌ |

**All configurations are slower than PyTorch on Nvidia.**

---

### 🔴 CRITICAL: Backend Failures (40%)

| Backend | Status | Error/Speedup |
|---------|--------|---------------|
| Nvidia | ✅ PASS | **0.705x** ❌ (slower) |
| **Tianshu** | ❌ FAIL | `RuntimeError: gemm of double is not supported on CoreX` |
| Muxi | ✅ PASS | 0.999x |
| **Ascend** | ❌ FAIL | `AssertionError: Tensor-likes are not close!` |
| **Hygon** | ✅ PASS | **13.957x** ⚠️ (anomaly) |

**Hygon anomaly**: Same algorithm gives **13.957x speedup** on Hygon but **0.705x on Nvidia**. This needs investigation.

---

### 🔴 CRITICAL: Sequential Kernel Design

```python
@triton.jit
def cholesky_kernel(A, L, N, stride_a, stride_l):
    pid = program_id(0)  # Each program handles ONE matrix
    
    for i in range(N):       # ← Sequential row loop
        for j in range(i + 1):  # ← Sequential column loop
            sum_val = tl.zeros((), dtype=tl.float32)
            if j > 0:
                for k in range(j):  # ← Sequential sum loop
                    sum_val = sum_val + tl.load(...) * tl.load(...)
            # Compute L[i,j]
```

**Why this is problematic**:

1. **Cholesky has data dependencies**: Row i depends on rows 0 to i-1
2. **Single program per matrix**: Low GPU utilization (batch_size programs total)
3. **O(N³) memory accesses**: Each element requires multiple loads/stores
4. **No parallelism**: Inner loops execute sequentially on one thread

**This defeats GPU parallelism** - Cholesky should use blocked algorithms.

---

### 🔴 Must Fix: `__all__` Sort Order (from Dongxu-H's review)

**File**: `src/flag_gems/ops/__init__.py`

```python
    "celu_",
+   "linalg_cholesky",  # ← WRONG! Should be in l section
    "clamp",
```

**Correct placement**:
```python
    "lift_fresh_copy",
+   "linalg_cholesky",  # ← Between lift_fresh_copy and linspace
    "linspace",
```

---

### 🟡 Suggestions from Previous Reviews

**1. Use Blocked Cholesky Algorithm**:

Reference: https://arxiv.org/abs/2009.01660

Blocked algorithm uses:
- Shared memory for panel updates
- Parallel panel factorization
- Reduced global memory traffic

**2. Add Missing Tests**:
- Batched matrices: `(2, 4, 4)`, `(3, 8, 8)`
- float64 dtype (PyTorch supports it)
- Edge cases: 1x1 matrix, near-singular matrices

**3. Handle Non-Positive-Definite Matrices**:

Current implementation silently modifies input:
```python
A_sym = (A + A.t().conj()) / 2  # Symmetrize
A_sym = A_sym + torch.eye(...) * eps  # Add eps to diagonal
```

This may produce incorrect results for non-PD matrices. Better approach:
- Detect non-PD during decomposition (negative sqrt)
- Raise `RuntimeError` with meaningful message

---

### 📋 Summary

| Issue | Severity | Status |
|-------|----------|--------|
| Performance regression (0.705x) | 🔴 Critical | 30% slower |
| Tianshu backend failure | 🔴 Critical | RuntimeError |
| Ascend backend failure | 🔴 Critical | AssertionError |
| `__all__` sort order | 🔴 Must Fix | Wrong position |
| Sequential kernel design | 🔴 Critical | Not GPU-optimized |
| Hygon anomaly (13.957x) | ⚠️ Investigation | Why so different? |

---

### 💡 Recommendations

#### Option A: Use cuSOLVER/cuBLAS Batched Cholesky

For Nvidia, use library functions:
```python
# cuSOLVER provides optimized batched Cholesky
torch.linalg.cholesky() already uses cuSOLVER internally
```

**This PR's Triton implementation is 30% slower** - consider using PyTorch's native as fallback.

#### Option B: Implement Blocked Algorithm

If Triton kernel is required, use blocked Cholesky:
1. Panel factorization (parallel)
2. Panel update (matrix-matrix multiply)
3. Use shared memory for panel

This requires significant redesign but can achieve competitive performance.

#### Option C: Keep for Specific Backends

Hygon shows 13.957x speedup - this implementation may work well on some architectures.
- Keep Triton kernel for Hygon
- Use PyTorch native for Nvidia
- Backend-specific overrides

---

### 🎯 Conclusion

**Request Changes** 🔴

**Critical Issues** (from previous reviews, still relevant):
1. **Performance**: 0.705x speedup (30% slower than PyTorch)
2. **Backend failures**: Tianshu and Ascend fail
3. **`__all__` sort order**: Must be fixed
4. **Algorithm**: Sequential kernel not GPU-optimized

**Required Actions**:
1. **Fix `__all__` sort order** - move to correct position
2. **Investigate Hygon anomaly** - why 13.957x vs 0.705x?
3. **Optimize kernel** - use blocked algorithm or fallback to cuSOLVER
4. **Fix/skip failing backends** - Tianshu, Ascend

**If performance cannot be improved on Nvidia**, consider using PyTorch's native `torch.linalg.cholesky()` as fallback, and keep Triton kernel only for backends where it shows benefit (Hygon).

File: src/flag_gems/ops/linalg_cholesky.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        raise ValueError("A must be a square matrix")
+
+    # Handle batch dimensions
+    if len(shape) == 2:
+        batch_size = 1
+    else:
+        batch_dims = shape[:-2]
+        batch_size = 1
+        for dim in batch_dims:
+            batch_size *= dim
```
Comment (Caeruleann) [inline src/flag_gems/ops/linalg_cholesky.py:94]: Please add 3D shape test case.

Comment (XDYuanzhuLee) [reply src/flag_gems/ops/linalg_cholesky.py:94]: Thanks. Added 3D batched shape test cases (2, 4, 4) and (3, 8, 8) in test_linalg_cholesky_batch.

**Comment by Dongxu-H** [comment]
LGTM

**Review by Dongxu-H — DISMISSED** [review]

File: tests/test_linalg_cholesky.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        + torch.eye(n, dtype=dtype, device=flag_gems.device) * 0.1
+    )
+
+    # For reference, convert to CPU and use torch.linalg.cholesky
+    ref_A = utils.to_reference(A)
+    ref_out = torch.linalg.cholesky(ref_A)
+
+    # For gems, use aten.cholesky with flag_gems
+    with flag_gems.use_gems():
+        res_out = torch.ops.aten.cholesky(A)
```
Comment (Caeruleann) [inline tests/test_linalg_cholesky.py]: Registered as linalg_cholesky, but the tests call torch.ops.aten.cholesky

File: tests/test_linalg_cholesky.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        + torch.eye(n, dtype=dtype, device=flag_gems.device) * 0.1
+    )
+
+    # For reference, convert to CPU and use torch.linalg.cholesky
+    ref_A = utils.to_reference(A)
+    ref_out = torch.linalg.cholesky(ref_A)
+
+    # For gems, use aten.linalg_cholesky with flag_gems
+    with flag_gems.use_gems():
+        res_out = torch.ops.aten.linalg_cholesky(A)
```
Comment (XDYuanzhuLee) [inline tests/test_linalg_cholesky.py:28]: Thanks for catching this. The test file already uses torch.ops.aten.linalg_cholesky matching the registered dispatch name. I also found the benchmark had the same mismatch - torch_op=torch.ops.aten.cholesky - and fixed it to torch.ops.aten.linalg_cholesky so the benchmarks gems path is correctly intercepted under flag_gems.use_gems().

**Review by Caeruleann — DISMISSED** [review]

**Review by Dongxu-H — CHANGES_REQUESTED** [review]
实现代码中声明支持 float32 和 float64：
   ```python
   assert A.dtype in (torch.float32, torch.float64), "linalg_cholesky only supports float32 and float64"
   ```
   
   但测试只覆盖了 float32，缺少 float64 的测试：
   ```python
   @pytest.mark.parametrize("dtype", [torch.float32])
   ```
   
   请添加 float64 测试用例。

**Review by Dongxu-H — CHANGES_REQUESTED** [review]
## 🔴 必须修复

1. **[性能] Speedup 0.705，慢 30%**
   
   算术平均 speedup 为 0.705，意味着比 PyTorch 慢约 30%。对于 Cholesky 分解这种计算密集型算子，性能显著低于 baseline 需要重点关注。
   
   | Size | Torch (ms) | Gems (ms) | Speedup |
   |------|------------|-----------|---------|
   | [128, 128] | 0.142 | 0.194 | 0.735 |
   | [256, 256] | 0.215 | 0.265 | 0.811 |
   
   建议：考虑优化 kernel（如使用分块算法、并行化列计算）以提升性能。

2. **[测试] float64 未测试**
   
   实现代码中声明支持 float32 和 float64：
   ```python
   assert A.dtype in (torch.float32, torch.float64), "linalg_cholesky only supports float32 and float64"
   ```
   
   但测试只覆盖了 float32，缺少 float64 的测试：
   ```python
   @pytest.mark.parametrize("dtype", [torch.float32])
   ```
   
   请添加 float64 测试用例。

---

**CI 状态**: python-op ✅ 通过

**配置排序**: operators.yaml ✅ 正确

**Backend 失败**: Tianshu/Ascend 失败可接受（主实现，Nvidia 通过）

**Logger 命名**: ✅ 正确

**Review by Dongxu-H — DISMISSED** [review]
benchmark 文件可添加 float64 dtype 测试（当前仅 float32）

**Review by Dongxu-H — APPROVED** [review]
LGTM

**Review by bin913 — APPROVED** [review]
